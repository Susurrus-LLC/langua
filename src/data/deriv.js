export const defData = {
  words: 5,
  type: 'any',
  lexemes: [
    {
      lexeme: 'nzi',
      definition: 'tree'
    },
    {
      lexeme: 'ilu',
      definition: 'flower'
    },
    {
      lexeme: 'ru',
      definition: 'person'
    },
    {
      lexeme: 'kwərə',
      definition: 'goat'
    },
    {
      lexeme: 'lɛtsi',
      definition: 'sword'
    }
  ],
  derivations: [
    {
      derivation: '-vo',
      gloss: 'person',
      definition:
        'Creates a noun referring to a person associated with the base.'
    },
    {
      derivation: '-lu',
      gloss: 'place',
      definition: 'Creates a noun denoting the location of the base.'
    },
    {
      derivation: 'ña-',
      gloss: 'abstract',
      definition: 'Creates an abstract noun.'
    },
    {
      derivation: 'k-sa',
      gloss: 'collective',
      definition: 'Creates a noun referring to a collection of bases.'
    }
  ],
  results: undefined
}
