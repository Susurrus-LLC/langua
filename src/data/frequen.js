export const defData = {
  corpus:
    'Nzɔ de’áxa Tsənaxa, ño lo sɛtsa’ɔk, ño lo de’áxa u Kasaga, ño lo mɛya ayru Tol o Əbe kwə, sa nzɔ xa mpaskale: Steyə rabɛ lu ñire lu mvunə poxe o lu pɛlɔ ah lo mɛwɛ axe, maxə ɔra latsi əbɔwə səyo.',
  consonants:
    'b/d/f/g/k/kw/l/m/mp/mv/n/nt/nts/nz/ñ/ŋk/ŋkw/p/r/s/t/ts/v/w/x,h/y/’',
  vowels: 'a,á/ɛ,ɛ́/e,é/ə,ə́/i,í/ɔ,ɔ́/o,ó/u,ú',
  distinguishCase: false,
  analyzed: false,
  results: undefined,
  hovered: undefined,
  filterSeg: 'all',
  filterSyll: 'all',
  filterWords: 'all'
}
