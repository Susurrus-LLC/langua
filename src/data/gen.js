export const defData = {
  subpatterns: [
    {
      selected: 'V',
      subpattern: 'a/e/i/o/u'
    },
    {
      selected: 'C',
      subpattern: 'p/t/k/b/d/g/s/m/n/l/r'
    },
    {
      selected: 'N',
      subpattern: 'm/n'
    }
  ],
  pattern: 'V/CV/CVN/CVCV/CVCVN',
  words: 100,
  newline: false,
  filterdupes: false
}

export const vars = [
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z'
]
