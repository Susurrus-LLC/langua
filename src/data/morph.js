export const defData = {
  categories: [
    'V=ieaou',
    'L=īēāōū',
    'C=ptcqbdgmnlrhs',
    'F=ie',
    'B=ou',
    'S=ptc',
    'Z=bdg'
  ],
  rewriteRules: ['lh=lj'],
  soundChanges: [
    '[sm]//_#',
    'i/j/_V',
    'L/V/_',
    'e//Vr_#',
    'v//V_V',
    'u/o/_#',
    'gn/nh/_',
    'S/Z/V_V',
    'c/i/F_t',
    'c/u/B_t',
    'p//V_t',
    'ii/i/_',
    'e//C_rV'
  ],
  lexicon: [
    'lector',
    'doctor',
    'focus',
    'jocus',
    'districtus',
    'cīvitatem',
    'adoptare',
    'opera',
    'secundus',
    'fīliam',
    'pōntem'
  ],
  outputFormat: 'oo',
  showDiff: true,
  showChanges: true,
  ruleReport: true,
  rewriteOutput: true,
  results: undefined
}
