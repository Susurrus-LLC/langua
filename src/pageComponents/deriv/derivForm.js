import React from 'react'
import PropTypes from 'prop-types'

import Form from '../../components/form'

import sharedStyles from '../../components/form/sharedForm.module.sass'

const DerivForm = () => <Form name='deriv-form' />

DerivForm.propTypes = {
  styles: PropTypes.string
}

export default DerivForm
