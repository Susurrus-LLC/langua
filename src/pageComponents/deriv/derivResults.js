import React from 'react'
import PropTypes from 'prop-types'

import Results from '../../components/results'

import sharedStyles from '../../components/results/sharedResults.module.sass'

const DerivResults = () => <Results />

DerivResults.propTypes = {
  styles: PropTypes.string
}

export default DerivResults
