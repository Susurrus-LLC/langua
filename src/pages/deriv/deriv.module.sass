@import ../../styles/variables.module

.words-input
  width: $ms6 * 1rem

.deriv-row, .adding-row
  width: 100%
  display: flex
  flex-direction: row
  align-items: baseline
  align-content: space-between

.deriv-row
  justify-content: space-between

.output-row
  justify-content: flex-start

.column-label
  margin-top: $ms0 * 1rem
  margin-bottom: 0

.lexeme-input, .word-label, .derivation-gloss-input, .gloss-label
  width: $ms5 * 2rem

.lexeme-def-input, .definition-label, .derivation-def-input
  flex-grow: 1

.derivation-input, .affix-label
  width: $ms4 * 2rem

.adding-row
  justify-content: flex-end
  padding: 1px 0
  margin: 1px 0

.output-box
  margin-right: $ms0 * 1rem
  margin-left: $ms0 * 1rem
  &:first-child
    margin-left: 0
  &:last-child
    margin-right: 0

.row-number
  width: $ms4 * 1rem

.term
  margin-bottom: 0
  font-weight: bold

.gloss
  margin-bottom: 0

.lex-gloss
  font-style: italic

.affix-gloss
  text-transform: lowercase
  font-variant: small-caps
  border-bottom: 1px dotted $blu
  transition: border-bottom 200ms

.affix-def
  color: lighten($black, 20%)
  font-size: $ms0 * 1rem
  visibility: hidden
  opacity: 0
  transition: opacity 200ms, visibility 0ms 200ms

.hovered-gloss
  border-bottom: 1px dotted $ong

.hovered-def
  visibility: visible
  opacity: 1
  transition: opacity 200ms, visibility 0ms 0ms
