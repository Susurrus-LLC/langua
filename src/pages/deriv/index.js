import React, { useState } from 'react'
import { graphql, useStaticQuery } from 'gatsby'

import Tool from '../../components/tool'

import DerivForm from '../../pageComponents/deriv/derivForm'
import DerivResults from '../../pageComponents/deriv/derivResults'
import derivService from '../../services/derivService'

import styles from './deriv.module.sass'

const DerivPage = () => {
  const data = useStaticQuery(graphql`
    query DerivToolQuery {
      site {
        siteMetadata {
          toolInfo {
            deriv {
              title
              link
              description
            }
          }
        }
      }
    }
  `)

  const metadata = data.site.siteMetadata
  const toolInfo = metadata.toolInfo.deriv

  // State

  const [words, setWords] = useState(derivService.getData().words)
  const [type, setType] = useState(derivService.getData().type)
  const [lexemes, setLexemes] = useState(derivService.getData().lexemes)
  const [derivations, setDerivations] = useState(
    derivService.getData().derivations
  )
  const [results, setResults] = useState(derivService.getData().results)

  // Functions

  const onChange = e => {
    const state = { words, type, lexemes, derivations, results }
    if (e.target.tagName === 'BUTTON') {
      e.preventDefault()
      if (e.target.type === 'submit') {
        // Derive the output
        const response = derivService.derive(state)
        state.results = response
        setResults(response)
        // Save the current state to storage
        derivService.setStorage(state)
      } else if (e.target.id === 'save') {
        // Save the current state to storage and generate a file
        derivService.save(state)
      } else if (e.target.id.slice(3) === 'Add') {
        const which = e.target.id.slice(0, 3)
        // When the add button is clicked, add a blank row to state
        const newData = derivService.add(state, which)
        setLexemes(newData.lexemes)
        setDerivations(newData.derivations)
      } else if (e.target.id.slice(1, 2) === 'c') {
        // When a row is cleared, delete teh corresponding row from state
        const idNum = e.target.id.slice(2)
        const which = e.target.id.slice(0, 1)
        const newData = derivService.clear(state, idNum, which)
        setLexemes(newData.lexemes)
        setDerivations(newData.derivations)
      }
    } else if (e.target.type === 'file') {
      e.preventDefault()
      // Open a file and parse it to restore saved state
      const file = e.target.files[0]
      const updateState = response => {
        if (response) {
          // Only change state if the file was successfully opened
          setWords(response.words)
          setType(response.type)
          setLexemes(response.lexemes)
          setDerivations(response.derivations)
          setResults(response.results)
        }
      }

      derivService.open(file, updateState)
    } else {
      const idNum = e.target.id.slice(2)
      const which = e.target.id.slice(0, 2)
      const val = e.target.value
      const name = e.target.name
      if (
        which === 'lw' ||
        which === 'ld' ||
        which === 'da' ||
        which === 'dl' ||
        which === 'dd'
      ) {
        const response = derivService.changeInput(state, idNum, which, val)
        if (response) {
          // Only change state if the input was successfully changed
          setWords(response.words)
          setType(response.type)
          setLexemes(response.lexemes)
          setDerivations(response.derivations)
          setResults(response.results)
        }
      } else if (e.target.name === 'words') {
        const response = derivService.wordNumChange(state, val)
        if (response) {
          // Only change state if the word number was successfully changed
          setWords(response.words)
          setType(response.type)
          setLexemes(response.lexemes)
          setDerivations(response.derivations)
          setResults(response.results)
        }
      } else {
        if (e.target.name === 'type') {
          setType(val)
        }
      }
    }
  }

  return (
    <Tool toolInfo={toolInfo}>
      <DerivForm
        styles={styles}
        data={{ words, type, lexemes, derivations }}
        change={onChange}
      />
      <DerivResults styles={styles} results={results} />
    </Tool>
  )
}

export default DerivPage
