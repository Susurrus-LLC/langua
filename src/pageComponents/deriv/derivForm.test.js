import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import DerivForm from './derivForm'
import { defData } from '../../data/deriv'

describe('DerivForm', () => {
  const styles = {
    derivRow: 'derivRow',
    lexemeInput: 'lexemeInput',
    lexemeDefInput: 'lexemeDefInput',
    clearButton: 'clearButton',
    derivRow: 'derivRow',
    derivationInput: 'derivationInput',
    derivationGlossInput: 'derivationGlossInput',
    derivationDefInput: 'derivationDefInput',
    addingRow: 'addingRow',
    lexemes: 'lexemes',
    columnLabel: 'columnLabel',
    wordLabel: 'wordLabel',
    definitionLabel: 'definitionLabel',
    derivations: 'derivations',
    affixLabel: 'affixLabel',
    glossLabel: 'glossLabel',
    wordsInput: 'wordsInput'
  }

  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(
      <DerivForm styles={styles} data={defData} change={jest.fn()} />,
      div
    )
    ReactDOM.unmountComponentAtNode(div)
  })

  it('renders accurately', () => {
    let tree = renderer
      .create(<DerivForm styles={styles} data={defData} change={jest.fn()} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
