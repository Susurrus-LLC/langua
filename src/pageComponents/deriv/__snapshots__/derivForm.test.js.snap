// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DerivForm renders accurately 1`] = `
<form
  className="form"
  name="deriv-form"
>
  <div
    className="lexemes part"
  >
    <h5
      className="sectionTitle"
    >
      Lexemes
    </h5>
    <div
      className="derivRow"
    >
      <h6
        className="columnLabel wordLabel"
      >
        Word
      </h6>
      <h6
        className="columnLabel definitionLabel"
      >
        Gloss
      </h6>
    </div>
    <div
      className="derivRow"
    >
      <input
        aria-label="Word for lexeme 0"
        className="lexemeInput"
        id="lw0"
        name="lw0"
        onChange={[MockFunction]}
        type="text"
        value="nzi"
      />
      <input
        aria-label="Gloss for lexeme 0"
        className="lexemeDefInput"
        id="ld0"
        name="ld0"
        onChange={[MockFunction]}
        type="text"
        value="tree"
      />
      <div
        className="clearButton"
      >
        <button
          className="btn danger small"
          id="lc0"
          onClick={[MockFunction]}
          role="button"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
    <div
      className="derivRow"
    >
      <input
        aria-label="Word for lexeme 1"
        className="lexemeInput"
        id="lw1"
        name="lw1"
        onChange={[MockFunction]}
        type="text"
        value="ilu"
      />
      <input
        aria-label="Gloss for lexeme 1"
        className="lexemeDefInput"
        id="ld1"
        name="ld1"
        onChange={[MockFunction]}
        type="text"
        value="flower"
      />
      <div
        className="clearButton"
      >
        <button
          className="btn danger small"
          id="lc1"
          onClick={[MockFunction]}
          role="button"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
    <div
      className="derivRow"
    >
      <input
        aria-label="Word for lexeme 2"
        className="lexemeInput"
        id="lw2"
        name="lw2"
        onChange={[MockFunction]}
        type="text"
        value="ru"
      />
      <input
        aria-label="Gloss for lexeme 2"
        className="lexemeDefInput"
        id="ld2"
        name="ld2"
        onChange={[MockFunction]}
        type="text"
        value="person"
      />
      <div
        className="clearButton"
      >
        <button
          className="btn danger small"
          id="lc2"
          onClick={[MockFunction]}
          role="button"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
    <div
      className="derivRow"
    >
      <input
        aria-label="Word for lexeme 3"
        className="lexemeInput"
        id="lw3"
        name="lw3"
        onChange={[MockFunction]}
        type="text"
        value="kwərə"
      />
      <input
        aria-label="Gloss for lexeme 3"
        className="lexemeDefInput"
        id="ld3"
        name="ld3"
        onChange={[MockFunction]}
        type="text"
        value="goat"
      />
      <div
        className="clearButton"
      >
        <button
          className="btn danger small"
          id="lc3"
          onClick={[MockFunction]}
          role="button"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
    <div
      className="derivRow"
    >
      <input
        aria-label="Word for lexeme 4"
        className="lexemeInput"
        id="lw4"
        name="lw4"
        onChange={[MockFunction]}
        type="text"
        value="lɛtsi"
      />
      <input
        aria-label="Gloss for lexeme 4"
        className="lexemeDefInput"
        id="ld4"
        name="ld4"
        onChange={[MockFunction]}
        type="text"
        value="sword"
      />
      <div
        className="clearButton"
      >
        <button
          className="btn danger small"
          id="lc4"
          onClick={[MockFunction]}
          role="button"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
    <div
      className="addingRow"
    >
      <button
        className="btn success small"
        id="lexAdd"
        onClick={[MockFunction]}
        role="button"
        type="button"
      >
        Add
      </button>
    </div>
  </div>
  <div
    className="derivations part"
  >
    <h5
      className="sectionTitle"
    >
      Derivational Affixes
    </h5>
    <div
      className="derivRow"
    >
      <h6
        className="columnLabel affixLabel"
      >
        Affix
      </h6>
      <h6
        className="columnLabel glossLabel"
      >
        Gloss
      </h6>
      <h6
        className="columnLabel definitionLabel"
      >
        Definition
      </h6>
    </div>
    <div
      className="derivRow"
    >
      <input
        aria-label="Affix 0"
        className="derivationInput"
        id="da0"
        name="da0"
        onChange={[MockFunction]}
        type="text"
        value="-vo"
      />
      <input
        aria-label="Gloss for affix 0"
        className="derivationGlossInput"
        id="dl0"
        name="dl0"
        onChange={[MockFunction]}
        type="text"
        value="person"
      />
      <input
        aria-label="Definition for affix 0"
        className="derivationDefInput"
        id="dd0"
        name="dd0"
        onChange={[MockFunction]}
        type="text"
        value="Creates a noun referring to a person associated with the base."
      />
      <div
        className="clearButton"
      >
        <button
          className="btn danger small"
          id="dc0"
          onClick={[MockFunction]}
          role="button"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
    <div
      className="derivRow"
    >
      <input
        aria-label="Affix 1"
        className="derivationInput"
        id="da1"
        name="da1"
        onChange={[MockFunction]}
        type="text"
        value="-lu"
      />
      <input
        aria-label="Gloss for affix 1"
        className="derivationGlossInput"
        id="dl1"
        name="dl1"
        onChange={[MockFunction]}
        type="text"
        value="place"
      />
      <input
        aria-label="Definition for affix 1"
        className="derivationDefInput"
        id="dd1"
        name="dd1"
        onChange={[MockFunction]}
        type="text"
        value="Creates a noun denoting the location of the base."
      />
      <div
        className="clearButton"
      >
        <button
          className="btn danger small"
          id="dc1"
          onClick={[MockFunction]}
          role="button"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
    <div
      className="derivRow"
    >
      <input
        aria-label="Affix 2"
        className="derivationInput"
        id="da2"
        name="da2"
        onChange={[MockFunction]}
        type="text"
        value="ña-"
      />
      <input
        aria-label="Gloss for affix 2"
        className="derivationGlossInput"
        id="dl2"
        name="dl2"
        onChange={[MockFunction]}
        type="text"
        value="abstract"
      />
      <input
        aria-label="Definition for affix 2"
        className="derivationDefInput"
        id="dd2"
        name="dd2"
        onChange={[MockFunction]}
        type="text"
        value="Creates an abstract noun."
      />
      <div
        className="clearButton"
      >
        <button
          className="btn danger small"
          id="dc2"
          onClick={[MockFunction]}
          role="button"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
    <div
      className="derivRow"
    >
      <input
        aria-label="Affix 3"
        className="derivationInput"
        id="da3"
        name="da3"
        onChange={[MockFunction]}
        type="text"
        value="k-sa"
      />
      <input
        aria-label="Gloss for affix 3"
        className="derivationGlossInput"
        id="dl3"
        name="dl3"
        onChange={[MockFunction]}
        type="text"
        value="collective"
      />
      <input
        aria-label="Definition for affix 3"
        className="derivationDefInput"
        id="dd3"
        name="dd3"
        onChange={[MockFunction]}
        type="text"
        value="Creates a noun referring to a collection of bases."
      />
      <div
        className="clearButton"
      >
        <button
          className="btn danger small"
          id="dc3"
          onClick={[MockFunction]}
          role="button"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
    <div
      className="addingRow"
    >
      <button
        className="btn success small"
        id="derAdd"
        onClick={[MockFunction]}
        role="button"
        type="button"
      >
        Add
      </button>
    </div>
  </div>
  <div
    className="control part"
  >
    <div
      className="controlSide controlLeft"
    >
      <div
        className="controlPiece"
      >
        <button
          className="btn neutral"
          onClick={[MockFunction]}
          role="button"
          type="submit"
        >
          Derive
        </button>
      </div>
      <div
        className="controlPiece"
      >
        <label
          htmlFor="words"
        >
          number of words to derive:
        </label>
        <input
          className="wordsInput"
          id="words"
          max="9999"
          min="1"
          name="words"
          onChange={[MockFunction]}
          type="number"
          value={5}
        />
      </div>
      <div
        className="controlPiece"
      >
        <label
          htmlFor="type"
        >
          output type:
        </label>
        <select
          id="type"
          name="type"
          onChange={[MockFunction]}
          value="any"
        >
          <option
            value="any"
          >
            any base, any derivation
          </option>
          <option
            value="single base"
          >
            single base
          </option>
          <option
            value="single derivation"
          >
            single derivation
          </option>
        </select>
      </div>
    </div>
    <div
      className="controlSide controlRight"
    >
      <div
        className="controlPiece"
      >
        <button
          className="btn success"
          id="save"
          onClick={[MockFunction]}
          role="button"
          type="button"
        >
          Save
        </button>
      </div>
      <div
        className="controlPiece"
      >
        <button
          className="btn"
          id="open"
          onClick={[Function]}
          role="button"
          type="button"
        >
          Open
        </button>
        <input
          className="hidden"
          id="file"
          name="file"
          onChange={[MockFunction]}
          type="file"
        />
      </div>
    </div>
  </div>
</form>
`;
